pyiron_workflow/job.py

```diff
@@ -0,0 +1,207 @@
+        _possible_ the same requirement from non-modal mode (importable nodes) will
+        apply.
+"""
+
+
+class NodeJob(TemplateJob):
+    __doc__ = (
+        """
+    This job is an intermediate feature for accessing pyiron's queue submission
+    infrastructure for nodes (function nodes, macros, or entire workflows).
+
+    It leans directly on the storage capabilities of the node itself, except for
+    the node class and name, and the storage backend mode, all of which are held in the
+    traditional job input. (Only the storage backend ever needs to be specified, the
+    node information gets populated automatically).
+
+    The job provides direct access to its owned node (as both input and output) on the
+    :attr:`node` attribute. The only requirement is that the node have an untouched
+    working directory (so we can make sure its files get stored _inside_ the job's
+    directory tree), and that it be compatible with the storage backend used.
+
+    Examples:
+        >>> from pyiron_base import Project
+        >>> from pyiron_workflow import Workflow
+        >>> import pyiron_workflow.job  # To get the job registered in JOB_CLASS_DICT
+        >>> 
+        >>> wf = Workflow("pyiron_node", overwrite_save=True)
+        >>> wf.answer = Workflow.create.standard.UserInput(42)  # Or your nodes
+        >>> 
+        >>> pr = Project("test")
+        >>> 
+        >>> nj = pr.create.job.NodeJob("my_node")
+        >>> nj.node = wf
+        >>> nj.run()  # doctest:+ELLIPSIS
+        The job my_node was saved and received the ID: ...
+        >>> print(nj.node.outputs.to_value_dict())
+        {'answer__user_input': 42}
+
+        >>> lj = pr.load(nj.job_name)
+        >>> print(nj.node.outputs.to_value_dict())
+        {'answer__user_input': 42}
+
+        >>> pr.remove_jobs(recursive=True, silently=True)
+        >>> pr.remove(enable=True)
+
+    """
+        + _WARNINGS_STRING
+    )
+
+    def __init__(self, project, job_name):
+        super().__init__(project, job_name)
+        self._python_only_job = True
+        self._write_work_dir_warnings = False
+        self._node = None
+        self.input._label = None
+        self.input._class_type = None
+        self.input.storage_backend = "h5io"  # Or "tinybase"
+
+    @property
+    def node(self) -> Node:
+        if self._node is None and self.status.finished:
+            self._load_node()
+        return self._node
+
+    @node.setter
+    def node(self, new_node: Node):
+        if self._node is not None:
+            raise ValueError("Node already set, make a new job")
+        elif self._node_working_directory_already_there(new_node):
+            self.raise_working_directory_error()
+        else:
+            self._node = new_node
+            self.input._class_type = (
+                f"{new_node.__class__.__module__}." f"{new_node.__class__.__name__}"
+            )
+            self.input._label = new_node.label
```

Comment:
[not a blocker for this PR]

my main concern/feedback is that there are many cases where a user may not necessarily want to create a new folder for every single step of their workflow. Am I correct in thinking that this will raise an error if this reuses a working dir from a node? In a lot of cases, a workflow will fail at some point, and then will it not be extremely clunky to reuse early-computations from the earlier successful workflow steps. 

This is mostly a philosophical thing, and need not be addressed in this PR.

Reply:
> a user may not necessarily want to create a new folder for every single step of their workflow.

In this case each `NodeJob` is related 1:1 with a `Node`; that may be a single `Function` node, or a whole complex `Workflow`. In this respect, running a `NodeJob` is guaranteed to make at least one directory (for the job) and populate at least one file (the node's HDF file). So far we're no different from a regular pyiron job in this respect. If the `Node` is a graph and some of its children produce files, they'll be nested inside the job's directory with a directory structure that mirrors the graph topology; this is normal for the `Composite` nodes.

> Am I correct in thinking that this will raise an error if this reuses a working dir from a node?

What it's doing is making sure that the node is both fresh (not reloaded) _and_ that it hasn't peeked at its working directory -- which would set it to `./{node_label}`, when we want to force it to be nested inside the job `./{job_name}/{node_label}`. So a workflow that has been saved previously will surely fail, some nodes may be able to be assigned post-execution, but not if they already peaked at their nodes, e.g. if it's some macro running Lammps and the Lammps node has written IO files to the node's directory.

> In a lot of cases, a workflow will fail at some point, and then will it not be extremely clunky to reuse early-computations from the earlier successful workflow steps.
>
> This is mostly a philosophical thing, and need not be addressed in this PR.

These last two I'll take together. Yes. Yep, definitely still clunky. In general the current implementation of file storage is not yet  prepared for restarting from failed nodes -- it's a run-and-done feature right now. I agree the issue shouldn't be blocking as right now I'd prioritize getting some version of this capability running to build familiarity with the pain points, but I am fully on board with making sure we get philosophical regularly as we go along!

Ultimately I see wrapping the nodes in a job as an intermediate step for database and queue-access functionality. Ideally, I'd like to see each node be able to make these connections individually, e.g. so you could have some `Workflow` where _only_ the `Vasp` node got shipped off to the queue.

This would involve more sophisticated treatment of the storage, e.g. having remotely run nodes serialize to their own HDF5 file, then merging this into the main root-most node's HDF5 file at reload time. Some more thoughts on this are [here](https://github.com/pyiron/pyiron_workflow/issues/182).

Similarly instead of re-running a node, a key feature for recovering partially executed graphs is to re-load instead of re-run output -- i.e. we should be checking if there is saved data, if the current input matches the saved input, and if so just reload the old output instead of re-running. Discussed [here](https://github.com/pyiron/pyiron_workflow/issues/169) but also not implemented yet.